api/agent/agent.go

```diff
@@ -540,16 +536,21 @@ func (s *hotSlot) exec(ctx context.Context, call *call) error {
 	// link the container id and id in the logs [for us!]
 	common.Logger(ctx).WithField("container_id", s.container.id).Info("starting call")
 
-	// swap in the new stderr logger & stat accumulator
-	oldStderr := s.container.swap(call.stderr, &call.Stats)
-	defer s.container.swap(oldStderr, nil) // once we're done, swap out in this scope to prevent races
+	// swap in fresh pipes & stat accumulator to not interlace with other calls that used this slot [and timed out]
+	stdinRead, stdinWrite := io.Pipe()
+	stdoutRead, stdoutWrite := io.Pipe()
+	defer stdinRead.Close()
+	defer stdoutWrite.Close()
+
+	proto := protocol.New(protocol.Protocol(call.Format), stdinWrite, stdoutRead)
+
+	swapBack := s.container.swap(stdinRead, stdoutWrite, call.stderr, &call.Stats)
+	defer swapBack() // NOTE: it's important this runs before the pipes are closed.
 
 	errApp := make(chan error, 1)
 	go func() {
-		// TODO make sure stdin / stdout not blocked if container dies or we leak goroutine
-		// we have to make sure this gets shut down or 2 threads will be reading/writing in/out
 		ci := protocol.NewCallInfo(call.Call, call.req)
-		errApp <- s.proto.Dispatch(ctx, ci, call.w)
+		errApp <- proto.Dispatch(ctx, ci, call.w)
```

Comment:
I wonder if now we can swap this code inside the go routine with a ctx.Done() select, which can close the proper pipe to unblock main go routine waiting on proto.Dispatch().

Reply:
If we invert this go routine, then no need to worry about thread safe headers.

Reply:
hmm, once we close the pipes though Dispatch could still be stuck on the Read from the request body (not a pipe) and will not get an error until that Read call returns to write it to the pipe writer. I could not find a way around this, tinkered with some ideas, seems one of the pains of io.Copy. for this case, we do need to set https://golang.org/pkg/net/http/#Server.ReadTimeout which would close it. for the pipe for stdout of the container, closing this is good enough to get it to bail, but the header race still exists I think since the context could timeout after parsing the stdout but while we're adding the headers (we have no mutex around this). 

> which can close the proper pipe to unblock main go routine waiting on proto.Dispatch().

i think this is happening, but i'm not sure if i am understanding correctly. we are closing the pipes once the `exec` select ticks (ctx.Done() is a case).

Reply:
Right... ReadTimeout on client would be an issue.. We can think about it later, especially if we are going to buffer req/resp, then client read timeout issue will be moved somewhere else (earlier), and then we can safely invert the go routine like;

```
go func() {
    select {
        case <-ctx.Done():
               // close pipe to unblock Dispatch()
        case err := <- s.errC:
              // close pipe to unblock Dispatch()
   }
}

// block here (stays on this go routine to avoid Header lock issue)
return s.proto.Dispatch(...)
```

Reply:
got it. yea it would be nice if we could get dispatch to bail correctly on the ctx, reading the request body is an issue whether we put into a buffer or not I think (dang io.Copy) wrt timing out. spent some time playing with the io copy thing and really didn't have a great solution, but if we yanked the request into a buffer in this thread we at least wouldn't have the header race (is it worth it?)

Reply:
I think we'll be buffering req/resp going forward, right? I do like this (despite increased mem footprint) since then we decouple these I/O (slow client keeping container busy, etc.) issues out of the picture.

Reply:
I know we'll buffer resp. But I think we should also buffer request.

Reply:
#326

Reply:
Oh, another gotcha for request buffering would be to wait until we have a container... (eg. we got resource token), otherwise we'd be piling up these in memory unnecessarily...

Reply:
good idea